core/core.go

```diff
@@ -411,18 +411,55 @@ func downloadBazelIfNecessary(version string, baseDirectory string, repos *Repos
 	}
 
 	destDir := filepath.Join(baseDirectory, pathSegment, "bin")
+	expectedSha256 := strings.ToLower(GetEnvOrConfig("BAZELISK_VERIFY_SHA256"))
+
+	tmpDestFile := "bazel-tmp" + platforms.DetermineExecutableFilenameSuffix()
 	destFile := "bazel" + platforms.DetermineExecutableFilenameSuffix()
 
 	destPath := filepath.Join(destDir, destFile)
 	if _, err := os.Stat(destPath); err == nil {
 		return destPath, nil
 	}
 
+	var tmpDestPath string
 	if url := GetEnvOrConfig(BaseURLEnv); url != "" {
-		return repos.DownloadFromBaseURL(url, version, destDir, destFile)
+		tmpDestPath, err = repos.DownloadFromBaseURL(url, version, destDir, tmpDestFile)
+	} else {
+		tmpDestPath, err = downloader(destDir, tmpDestFile)
+	}
+	if err != nil {
+		return "", err
+	}
+
+	if len(expectedSha256) > 0 {
+		f, err := os.Open(tmpDestPath)
+		if err != nil {
+			os.Remove(tmpDestPath)
```

Comment:
What about defer'ing os.Remove(tmpDestPath) and f.Close()?

Reply:
Sure, switched to defer for the file deletion.

But I cannot do this for Close. That was what I originally did, but then encountered failures on Windows because keeping the file open interferes with the rename. Added a comment.